Diverse witnesses in shelley

paweljakubas opened the pull request:
# Issue Number

<!-- Put here a reference to the issue this PR relates to and which requirements it tackles -->
#1844 

# Overview

<!-- Detail in a few bullet points the work accomplished in this PR -->

- [x] I have added WitnessTag
- [x] I have added impl of mkTx for Byron/Icarus wtinesses
- [x] I have added unit tests verifying decodeTx for Byron/Icarus witnesses
- [x] I have updated fee estimation
- [x] I have added integration tests illustrating the case 


# Comments

<!-- Additional comments or screenshots to attach if any -->

 
<!-- 
Don't forget to:

 ✓ Self-review your changes to make sure nothing unexpected slipped through
 ✓ Assign yourself to the PR
 ✓ Assign one or several reviewer(s)
 ✓ Once created, link this PR to its corresponding ticket
 ✓ Assign the PR to a corresponding milestone
 ✓ Acknowledge any changes required to the Wiki
-->

KtorZ commented on lib/shelley/test/unit/Cardano/Wallet/Shelley/TransactionSpec.hs:
```diff
@@ -298,3 +382,15 @@ dummyAddress = Address $ BS.pack $ 1 : replicate 64 0
+    let n = fromIntegral (natVal $ Proxy @(EntropySize mw)) `div` 8
+    bytes <- BS.pack <$> vector n
+    let ent = unsafeMkEntropy @(EntropySize mw) bytes
+    return $ entropyToMnemonic ent
```
Why not simply use `genEntropy` with `EntropySize mw` to get the right number of bytes?

KtorZ commented on lib/shelley/src/Cardano/Wallet/Shelley/Transaction.hs:
```diff
@@ -355,6 +384,26 @@ _decodeSignedTx bytes = do
+        pure $ mkByronWitness pm sigData ks
+    pure
+        ( Tx (Hash sigData) (second coin <$> ownedIns) outs
+        , SealedTx $ CBOR.toStrictByteString $ CBOR.encodeSignedTx tx witnesses
```
I don't think this is right. This is how transaction were serialized in Byron, but in Shelley, they are following a different scheme. Therefore, we should serialize transaction using the new shelley methods; only the witnesses are different!

paweljakubas committed:
add tx witness tag

paweljakubas committed:
mkWitness -> mkShelleyWitness

paweljakubas committed:
introduce mkByronWitness

paweljakubas committed:
specialize witness creation in mkTx depending on witness tag

paweljakubas committed:
add test in TransactionSpec for byron witness

Commit:
Regenerate nix

paweljakubas committed:
sneaking witness tag into _minimumFee and _estimateMaximumNumberOfInputs

paweljakubas committed:
dependant computeTxSize on witness size

paweljakubas committed:
make Byron witness rountrip in TransactionSpec working

paweljakubas commented on lib/shelley/src/Cardano/Wallet/Shelley/Transaction.hs:
```diff
@@ -164,25 +194,34 @@ mkTx
+                (k, pwd) <- lookupPrivateKey keyFrom addr
+                pure $ mkByronWitness unsigned pm (getRawKey k, pwd)
+            pure $ SL.WitnessSet mempty mempty bootstrapWits
+                <> mkExtraWits unsigned --- TO_DO - this is needed?
```
this one is maybe not needed - I do not know at this moment

Anviking replied:
Yes, if extra witnesses are needed for certificates or whatnot, they end up here

paweljakubas replied:
yeah, but in byron transactions it should not happen to my best knowledge, right?

paweljakubas committed:
make shelley integration main compile

KtorZ commented on lib/shelley/src/Cardano/Wallet/Shelley/Transaction.hs:
```diff
@@ -421,7 +468,32 @@ computeTxSize proxy action cs =
+          where
+            credential = Hash.UnsafeHash $ L8.toStrict chaff
+            addrAttr = Byron.mkAttributes $ Byron.AddrAttributes
+                       (Just $ Byron.HDAddressPayload (bloatChaff keyLen))
```
Why `keyLen` here :thinking:  ? The HDAddressPayload is a ChaChaPoly digest + nonce that totals 28 bytes + 2 bytes because it's CBOR-encoded. I believe the `keyLen` is 32 bytes here so it's probably not a bit deal, but that is quite misleading and error-prone.

More importantly, there's only a HDPayload for /Byron/ type of addresses, Icarus do not.

paweljakubas replied:
addressed here https://github.com/input-output-hk/cardano-wallet/pull/1849/commits/1209bca45f0c0dd9cd8615355eee19df11893e46

KtorZ commented on lib/shelley/src/Cardano/Wallet/Shelley/Transaction.hs:
```diff
@@ -421,7 +468,32 @@ computeTxSize proxy action cs =
+            credential = Hash.UnsafeHash $ L8.toStrict chaff
+            addrAttr = Byron.mkAttributes $ Byron.AddrAttributes
+                       (Just $ Byron.HDAddressPayload (bloatChaff keyLen))
+                       (toByronNetworkMagic mainnetMagic)
```
mainnet is hard-coded here which is sort of wrong because mainnet addresses don't embed any discrimination attribute whereas testnet address do. So there's a significant size variation between the two.

KtorZ replied:
suggestion: rely on the `proxy` above to know whether mainnet or testnet is needed. See also: `fromNetworkDiscriminant`

paweljakubas replied:
can I retrieve protocolMagic easily from proxy? :thinking: 
I have above (I agree it was quick hack) : 
```
763 toByronNetworkMagic :: W.ProtocolMagic -> Byron.NetworkMagic                                                                                                                                                   
764 toByronNetworkMagic pm@(W.ProtocolMagic magic) =                                                                                                                                                               
765     if pm == W.mainnetMagic then                                                                                                                                                                               
766         Byron.NetworkMainOrStage                                                                                                                                                                               
767     else                                                                                                                                                                                                       
768         Byron.NetworkTestnet (fromIntegral magic)
```

KtorZ replied:
In theory they should match indeed.

KtorZ commented on lib/shelley/src/Cardano/Wallet/Shelley/Transaction.hs:
```diff
@@ -504,6 +576,20 @@ unsafeMkEd25519 :: XPub -> Ed25519.PublicKey
+    let signingKey = Crypto.SigningKey prv
+        bytes = xprvToBytes prv
+        addrAttr = Byron.mkAttributes $ Byron.AddrAttributes
+            (Just $ Byron.HDAddressPayload bytes)
```
I don't think we should send the user's private key over in a witness! 
The HDAddressPayload corresponds to the encrypted derivation path that is present only on Byron-Random addresses. I assume that what we needs here is: 

a) To pass down the corresponding address to `mkByronWitness`
b) Have a function in the Compatibility module to convert an `Address` into a `AddrAttributes`. We have all the information we need in the Address already.

paweljakubas replied:
have a try to deliver proper solution in https://github.com/input-output-hk/cardano-wallet/pull/1849/commits/72c70d0f90a40ac481bdbebd48199d67f0745163

KtorZ reviewed (changes requested):
The witness for Byron needs to be corrected, in particular the HD payload needs to be inferred from the address!

paweljakubas committed:
add integration test with byron witnesses

paweljakubas committed:
add toHDAddressPayload

paweljakubas committed:
correct partially dummy values

Anviking commented on lib/shelley/src/Cardano/Wallet/Shelley/Transaction.hs:
```diff
@@ -164,25 +195,34 @@ mkTx
+                (k, pwd) <- lookupPrivateKey keyFrom addr
+                pure $ mkShelleyWitness unsigned (getRawKey k, pwd)
+
+            let withdrawalsWits
```
I was first wondering: why not factor out `fmap Set.fromList $ forM (CS.inputs cs) $ \(_, TxOut addr _)` from both cases, and just differentiate on the witness construction?

Then I see `withdrawalsWits` are only in the `TxWitnessShelleyUTxO` case. Wouldn't it _technically_ be possible (at least for people using cardano-wallet as a library) to spend both byron funds and withdrawals? Would it not be as well to construct `withdrawalsWits` in both cases then?

paweljakubas replied:
that's good question - I have also this in the back of my head. I planned to have working byron/icarus tx...and then in the next PR check this possibility

KtorZ replied:
> to spend both byron funds and withdrawals?

Technically yes, but not with the wallet, because there are no withdrawals coming with our definition of Byron wallets.

KtorZ committed:
add Byron & Icarus faucet wallets to Shelley genesis

KtorZ committed:
sign Byron transaction with unencrypted private keys!

KtorZ committed:
make sure to unwrap the nested HDPayload when deconstructing Byron addresses

KtorZ committed:
do not create actual Byron witness when estimating

Doing to crypto takes time here, and the coin selection algorithm do estimate fee quite many times in the process. So instead, it's better to use dummy values.

KtorZ committed:
remove traces and fix fee estimation for Byron wallets

  The context fee estimator is showing its limit. It has actually
  become quite annoying to maintain two separate fee estimator.

  We should stop using this fee estimator and rely on the API instead.
  For now, I am manually adding margins necessary to account for Byron
  witnesses that are more expensive. Min and Max depends on whether we
  are making the transaction from an `Icarus` or `Random`.

  Fortunately, the API estimates this correctly!

paweljakubas committed:
remove feeEstimator and use fee endpoint

paweljakubas committed:
address protocolMagic in computeTxSize

paweljakubas committed:
add tests for both testnet and mainnet

KtorZ commented on lib/core-integration/src/Test/Integration/Scenario/API/Shelley/Transactions.hs:
```diff
@@ -464,6 +465,70 @@ spec = do
+                (Link.getWallet @'Byron wByron) Default Empty
+            expectField
+                (#balance . #available)
+                (`shouldBe` Quantity (faucetAmt - feeEstMax - amt)) ra2
```
:+1:

KtorZ reviewed (approved):
:+1:

paweljakubas committed:
test Icarus wallet
